internal/pkg/peer/blocksprovider/bft_header_receiver.go

```diff
@@ -0,0 +1,164 @@
+	lastHeaderOK       bool
+	logger             *flogging.FabricLogger
+}
+
+func NewBFTHeaderReceiver(
+	chainID string,
+	endpoint string,
+	client orderer.AtomicBroadcast_DeliverClient,
+	msgVerifier BlockVerifier,
+	logger *flogging.FabricLogger,
+) *bftHeaderReceiver {
+	hRcv := &bftHeaderReceiver{
+		chainID:           chainID,
+		stopChan:          make(chan struct{}, 1),
+		endpoint:          endpoint,
+		client:            client,
+		msgCryptoVerifier: msgVerifier,
+		logger:            logger.With("endpoint", endpoint),
+	}
+	return hRcv
+}
+
+// DeliverHeaders starts to deliver headers from the stream client
+func (hr *bftHeaderReceiver) DeliverHeaders() {
+	defer func() {
+		_ = hr.CloseSend()
+	}()
+
+	hr.logger.Debugf("[%s] Starting to deliver headers", hr.chainID)
+	hr.setStarted()
+
+	for !hr.IsStopped() {
+		msg, err := hr.client.Recv()
+		if err != nil {
+			hr.logger.Warningf("[%s] Receive error: %s", hr.chainID, err.Error())
+			return
+		}
+
+		switch t := msg.GetType().(type) {
+		case *orderer.DeliverResponse_Status:
+			if t.Status == common.Status_SUCCESS {
+				hr.logger.Warningf("[%s] Warning! Received %s for a seek that should never complete", hr.chainID, t.Status)
+				return
+			}
+
+			hr.logger.Errorf("[%s] Got bad status %s", hr.chainID, t.Status)
+			return
+
+		case *orderer.DeliverResponse_Block:
+			blockNum := t.Block.Header.Number
+
+			// do not verify, just save for later, in case the block-receiver is suspected of censorship
+			hr.logger.Debugf("[%s][%s] Saving block with header & metadata, blockNum = [%d], block = [%v]", hr.chainID, hr.endpoint, blockNum, t.Block)
+			hr.mutex.Lock()
+			hr.lastHeader = t.Block
+			hr.lastHeaderTime = time.Now()
+			hr.lastHeaderVerified = false
+			hr.lastHeaderOK = false
+			hr.mutex.Unlock()
+
+		default:
+			hr.logger.Warningf("[%s] Received unknown response type: %v", hr.chainID, t)
+			return
+		}
+	}
+
+	hr.logger.Debugf("[%s] Stopped to deliver headers from endpoint: %s", hr.chainID, hr.endpoint)
+}
+
+func (hr *bftHeaderReceiver) IsStopped() bool {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	return hr.stop
+}
+
+func (hr *bftHeaderReceiver) IsStarted() bool {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	return hr.started
+}
+
+func (hr *bftHeaderReceiver) setStarted() {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	hr.started = true
+}
+
+// CloseSend closes the client connection
+func (hr *bftHeaderReceiver) CloseSend() error {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+
+	if hr.stop {
+		return nil
+	}
+
+	hr.stop = true
+	_ = hr.client.CloseSend()
+	// TODO close the underlying connection as well
+	close(hr.stopChan)
+
+	return nil
+}
+
+// LastBlockNum returns the last block number
+func (hr *bftHeaderReceiver) LastBlockNum() (uint64, time.Time, error) {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+
+	if hr.lastHeader == nil {
+		return 0, time.Unix(0, 0), errors.New("Not found")
+	}
+
+	if !hr.lastHeaderVerified {
+		hr.lastHeaderVerified = true
```

Comment:
you are setting this to true prematurely, no? what if it's not verified in the end? don't you need to set it to false then?

Reply:
when we receive a header we do 
```
// do not verify, just save for later, in case the block-receiver is suspected of censorship
			hr.logger.Debugf("[%s][%s] Saving block with header & metadata, blockNum = [%d], block = [%v]", hr.chainID, hr.endpoint, blockNum, t.Block)
			hr.mutex.Lock()
			hr.lastHeader = t.Block
			hr.lastHeaderTime = time.Now()
			hr.lastHeaderVerified = false
			hr.lastHeaderOK = false
			hr.mutex.Unlock()
```